core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

```diff
@@ -623,15 +623,13 @@ private[spark] object SparkHadoopUtil extends Logging {
       fs.create(path)
     } else {
       try {
-        // Use reflection as this uses APIs only available in Hadoop 3
-        val builderMethod = fs.getClass().getMethod("createFile", classOf[Path])
         // the builder api does not resolve relative paths, nor does it create parent dirs, while
         // the old api does.
         if (!fs.mkdirs(path.getParent())) {
           throw new IOException(s"Failed to create parents of $path")
         }
         val qualifiedPath = fs.makeQualified(path)
-        val builder = builderMethod.invoke(fs, qualifiedPath)
+        val builder = fs.createFile(qualifiedPath)
         val builderCls = builder.getClass()
         // this may throw a NoSuchMethodException if the path is not on hdfs
         val replicateMethod = builderCls.getMethod("replicate")
```

Comment:
I originally intended to modify it as follows:
```scala
 fs.createFile(qualifiedPath) match {
    case hb: HdfsDataOutputStreamBuilder => hb.replicate().build()
    case _ => fs.create(path)
 }
```
But I'm not sure if there are other types of `FSDataOutputStreamBuilder` that support the `replicate` method. Do you have any better suggestions? @pan3793 @sunchao

Reply:
I'm +1 on the existing reflection approach since it can support any implementation which supports the `replicate` method.

Reply:
friendly ping @sunchao

Reply:
I think this method is very private and only available for `HdfsDataOutputStreamBuilder`. I'm OK if it stay as it is.

Reply:
QQ: If this is specific to `HdfsDataOutputStreamBuilder` and we are expecting to invoke `replicate` on that specific class, why not tighten the code to do it specifically for that class (and its subclasses) ?
With hadoop 2.x, I can understand the generic replicate method invocation - but we have the opportunity to be more specific now.

I am fine with move that snippet to a different PR though.

Reply:
> I think this method is very private and only available for `HdfsDataOutputStreamBuilder`. I'm OK if it stay as it is.

Can we confirm this?

My concern is if there is a `XDataOutputStreamBuilder`(Maybe user specific implementation),  if `XDataOutputStreamBuilder` is subclass of `FSDataOutputStreamBuilder ` instead of `HdfsDataOutputStreamBuilder`, but `XDataOutputStreamBuilder` also happens to have `replicate ` method. 

In this scenario, the current way can be executed, but changing to similar to 

```
 fs.createFile(qualifiedPath) match {
    case hb: HdfsDataOutputStreamBuilder => hb.replicate().build()
    case _ => fs.create(path)
 }
```
will inadvertently disrupt this behavior, so I let it stay as it is in this pr ...

Reply:
The intent here is not to invoke any random `replicate` method which might be sitting inside the `FSDataOutputStreamBuilder`  - but to do so specifically for hdfs `HdfsDataOutputStreamBuilder`/subclasses.
Unfortunately, this is not yet resolved (see HDFS-14038).

See [this discussion](https://github.com/apache/spark/pull/22881#discussion_r229102457) for more.

As part of this fix, if there are other implementations which get broken due to this change, we should address them appropriately on a case by case basis or get hdfs team to fix their interfaces : not keep a broken contract.

Reply:
In other words, I would argue to make the change proposed and remove the reflection:
```
fs.createFile(qualifiedPath) match {
    case hb: HdfsDataOutputStreamBuilder => hb.replicate().build()
    case _ => fs.create(path)
 }
```

As I mentioned before, we can move that to a separate PR though - to derisk this specific PR (in case we do find credible reasons to revert/fix the change)

Reply:
> As I mentioned before, we can move that to a separate PR though - to derisk this specific PR (in case we do find credible reasons to revert/fix the change)

Agreed

Reply:
that makes sense, thanks @mridulm for detailed explanation

Reply:
+1 on removing the reflection

Reply:
hmm... I still want to put it in an independent pr

Reply:
> +1 on removing the reflection

testing, will give a new one later